src/SlidingSyncManager.ts

```diff
@@ -0,0 +1,240 @@
+                required_state: [
+                    [EventType.RoomJoinRules, ""], // the public icon on the room list
+                    [EventType.RoomAvatar, ""], // any room avatar
+                    [EventType.RoomTombstone, ""], // lets JS SDK hide rooms which are dead
+                    [EventType.RoomEncryption, ""], // lets rooms be configured for E2EE correctly
+                    [EventType.RoomCreate, ""], // for isSpaceRoom checks
+                    [EventType.RoomMember, this.client.getUserId()], // lets the client calculate that we are in fact in the room
+                ],
+            };
+            list = Object.assign(list, updateArgs);
+        } else {
+            const updatedList = Object.assign({}, list, updateArgs);
+            // cannot use objectHasDiff as we need to do deep diff checking
+            if (JSON.stringify(list) === JSON.stringify(updatedList)) {
+                logger.debug("list matches, not sending, update => ", updateArgs);
+                return list;
+            }
+            list = updatedList;
+        }
+
+        try {
+            // if we only have range changes then call a different function so we don't nuke the list from before
+            if (updateArgs.ranges && Object.keys(updateArgs).length === 1) {
+                await this.slidingSync.setListRanges(listIndex, updateArgs.ranges);
+            } else {
+                await this.slidingSync.setList(listIndex, list);
+            }
+        } catch (err) {
+            logger.debug("ensureListRegistered: update failed txn_id=", err);
+        }
+        return this.slidingSync.getList(listIndex);
+    }
+
+    async setRoomVisible(roomId: string, visible: boolean): Promise<string> {
+        await this.configurePromise;
+        const subscriptions = this.slidingSync.getRoomSubscriptions();
+        if (visible) {
+            subscriptions.add(roomId);
+        } else {
+            subscriptions.delete(roomId);
+        }
+        logger.log("SlidingSync setRoomVisible:", roomId, visible);
+        this.slidingSync.modifyRoomSubscriptions(subscriptions);
+
+        return new Promise((resolve) => {
+            if (this.client.getRoom(roomId)) {
+                resolve(roomId); // we have data already for this room, show immediately e.g it's in a list
+                return;
+            }
+            const resolveOnSubscribed = (gotRoomId: string) => {
+                if (roomId === gotRoomId) {
+                    // we processed a /sync response which returned this subscription
+                    this.slidingSync.off(SlidingSyncEvent.RoomData, resolveOnSubscribed);
+                    resolve(roomId);
+                }
+            };
+            // wait until the next sync before returning as RoomView may need to know the current state
+            this.slidingSync.on(SlidingSyncEvent.RoomData, resolveOnSubscribed);
```

Comment:
If I am nervously clicking around in the room list :laughing: can this be an issue? For instance a listener is registered and then the subscription is deleted. Will the event still arrive, so that it can be unregistered?

Reply:
No, it won't actually. This pattern exists from before sliding sync had transaction IDs, I should be using that instead, which will reject under this circumstance. Thanks!